pkg/util/merger.go

```diff
@@ -0,0 +1,50 @@
+package util
+
+import "github.com/prometheus/common/model"
+
+// MergeSampleSets merges and dedupes two sets of already sorted sample pairs.
+func MergeSampleSets(a, b []model.SamplePair) []model.SamplePair {
+	result := make([]model.SamplePair, 0, len(a)+len(b))
+	i, j := 0, 0
+	for i < len(a) && j < len(b) {
+		if a[i].Timestamp < b[j].Timestamp {
+			result = append(result, a[i])
+			i++
+		} else if a[i].Timestamp > b[j].Timestamp {
+			result = append(result, b[j])
+			j++
+		} else {
+			result = append(result, a[i])
+			i++
+			j++
+		}
+	}
+	for ; i < len(a); i++ {
+		result = append(result, a[i])
+	}
+	for ; j < len(b); j++ {
+		result = append(result, b[j])
+	}
+	return result
+}
+
+// MergeNSampleSets merges and dedupes n sets of already sorted sample pairs.
+func MergeNSampleSets(sampleSets ...[]model.SamplePair) []model.SamplePair {
```

Comment:
In case this could be useful, here is an example of doing the n-way merge using a "dedup iterator" on a "merging iterator" on a collection of "iterators": https://play.golang.org/p/5d5F7EKgZo
Lack of generics make this pretty ugly, but feedback welcome. 
The thing about it is that it is nicely composable.